event/system.go

```diff
@@ -0,0 +1,145 @@
+package event
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Subscriber is a limited interface of System that only allows subscribing and
+// unsubscribing.
+type Subscriber interface {
+	Subscribe(events ...Type) (subID uint64, eventsCh <-chan *Event)
+	Unsubscribe(subID uint64)
+}
+
+// System keeps track of subscribers, and allows subscribing to and emitting
+// events.
+type System struct {
+	subMx       sync.RWMutex
+	subIDCount  uint64
+	subscribers map[Type]map[uint64]chan *Event
+	eventBuffer int
+	logger      logrus.FieldLogger
+}
+
+// NewEventSystem returns a new System.
+// eventBuffer determines the size of the Event channel buffer. Events might be
+// dropped if this buffer is full and there are no event listeners, or if events
+// are emitted very quickly and the event handler goroutine is busy. It is
+// recommended to handle events in a separate goroutine to not block the
+// listener goroutine.
+func NewEventSystem(eventBuffer int, logger logrus.FieldLogger) *System {
+	return &System{
+		subscribers: make(map[Type]map[uint64]chan *Event),
+		eventBuffer: eventBuffer,
+		logger:      logger, // TODO: Log some debug events
+	}
+}
+
+// Subscribe to one or more events. It returns a subscriber ID that can be
+// used to unsubscribe, and an Event channel to receive events.
+// It panics if events is empty.
+func (s *System) Subscribe(events ...Type) (subID uint64, eventsCh <-chan *Event) {
+	if len(events) == 0 {
+		panic("must subscribe to at least 1 event type")
+	}
+	s.subMx.Lock()
+	defer s.subMx.Unlock()
+	s.subIDCount++
+	subID = s.subIDCount
+	evtCh := make(chan *Event, s.eventBuffer)
+	for _, evt := range events {
+		if s.subscribers[evt] == nil {
+			s.subscribers[evt] = make(map[uint64]chan *Event)
+		}
+		s.subscribers[evt][subID] = evtCh
+	}
+	return subID, evtCh
+}
+
+// Emit the event to all subscribers of its type.
+// It returns a function that can be optionally used to wait for all subscribers
+// to process the event (by signalling via the Done method).
+func (s *System) Emit(event *Event) (wait func(context.Context) error) {
+	s.subMx.RLock()
+	defer s.subMx.RUnlock()
+	totalSubs := len(s.subscribers[event.Type])
+	if totalSubs == 0 {
+		return func(context.Context) error { return nil }
+	}
+
+	if event.Done == nil {
+		event.Done = func() {}
+	}
+	origDoneFn := event.Done
+	doneCh := make(chan struct{}, s.eventBuffer)
+	doneFn := func() {
+		origDoneFn()
+		select {
+		case doneCh <- struct{}{}:
+		default:
+		}
+	}
+	event.Done = doneFn
+
+	for _, evtCh := range s.subscribers[event.Type] {
+		select {
+		case evtCh <- event:
+		default:
+		}
+	}
+
+	return func(ctx context.Context) error {
+		var doneCount int
+		for {
+			if doneCount == totalSubs {
+				close(doneCh)
+				return nil
+			}
+			select {
+			case <-doneCh:
+				doneCount++
+			case <-ctx.Done():
```

Comment:
:thinking: since we derive from the `lingerCtx` could it be a constant error when somebody uses `Ctrl + C` un case of test end, I remember that the previous version of the code just returns without error when `ctx.Done`

Reply:
Yeah, you're right.

I wasn't sure whether `lingerCtx` is the correct base in this case, and we can't use `runCtx` since that can be aborted before waiting for events to be processed. I changed the base to `globalCtx` in c8b17b83b, and that works fine from my tests with `--linger` (it doesn't return an error). I think this should depend on a higher-level context, and not an entirely separate one, to ensure it can exit cleanly if needed.

As for returning without an error, I think that should only be done if we receive the signal that all events were processed. Previously it would silently exit if the context is done, but it should've raised a warning instead.

Now we don't have that distinction since we reuse the context timeout itself, but I think this is simpler, and we don't to handle a separate timeout.